### PR TITLE
executor: Add orphaned VM janitor

### PR DESCRIPTION
**Pull request by efritz**
Fixes https://github.com/sourcegraph/sourcegraph/issues/23356.

Follow-up PR will add additional dashboards to display new metrics.

### CODE REVIEW

File: enterprise/cmd/executor/internal/worker/handler.go [hunk 1]
```diff
@@ -92,13 +94,24 @@ func (h *handler) Handle(ctx context.Context, record workerutil.Record) (err err
...
 		return err
 	}
 
+	// Construct a unique name for the VM prefixed by something that differentiates
+	// VMs created by this executor instance and another one that happens to run on
+	// the same host (as is the case in dev).
+	name := fmt.Sprintf("%s-%s", h.options.VMPrefix, vmNameSuffix.String())
+
+	// Before we setup a VM (and after we teardown), mark the name as in-use so that
+	// the janitor process cleaning up orphaned VMs doesn't try to stop/remvoe the one
```
Comment (eseliger) [inline enterprise/cmd/executor/internal/worker/handler.go:108]:
```suggestion
	// the janitor process cleaning up orphaned VMs doesn't try to stop/remove the one
```

File: enterprise/cmd/executor/internal/janitor/orphaned_vms.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		} else {
+			j.metrics.numVMsRemoved.Inc()
+		}
+	}
+
+	return nil
+}
+
+func (j *orphanedVMJanitor) HandleError(err error) {
+	j.metrics.numErrors.Inc()
```
Comment (eseliger) [inline enterprise/cmd/executor/internal/janitor/orphaned_vms.go:71]: Should we do a withlabel and pass the prefix? Or hostname or whatever, so they can be distinguished in any way

File: enterprise/cmd/executor/internal/janitor/orphaned_vms.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+func (j *orphanedVMJanitor) Handle(ctx context.Context) (err error) {
+	runningVMs, err := currentlyRunningVMs(ctx, j.prefix)
+	if err != nil {
+		return err
+	}
+
+	for _, id := range findOrphanedVMs(runningVMs, j.names.Slice()) {
+		log15.Info("Removing orphaned vm", "id", id, "error", err)
+
+		if removeErr := exec.CommandContext(ctx, "ignite", "rm", "-f", id).Run(); removeErr != nil {
```
Comment (eseliger) [inline enterprise/cmd/executor/internal/janitor/orphaned_vms.go:60]: Does rm -f also stop it?

**Review by eseliger — APPROVED** [review]
Go go go 🚀

File: enterprise/cmd/executor/internal/janitor/orphaned_vms.go (hunk 2 above)
Comment (efritz) [reply enterprise/cmd/executor/internal/janitor/orphaned_vms.go:71]: The prometheus service discovery config should handle that and group them by instance.

File: enterprise/cmd/executor/internal/janitor/orphaned_vms.go (hunk 3 above)
Comment (efritz) [reply enterprise/cmd/executor/internal/janitor/orphaned_vms.go:60]: It _should_, but `ignite stop` and `ignite rm -f` should also have worked with these exact arguments earlier, but we see those fail.

**Pull request merged**